src/subscript-loc.c

```diff
@@ -152,35 +152,47 @@ static SEXP int_filter_zero(SEXP subscript, R_len_t n_zero) {
 // From compare.c
 int qsort_icmp(const void* x, const void* y);
 
-static void int_check_consecutive(SEXP subscript, R_len_t n,
+static void int_check_consecutive(SEXP subscript, R_len_t n, R_len_t n_extend,
                                   const struct location_opts* opts) {
-  SEXP sorted = PROTECT(Rf_duplicate(subscript));
-  int* p_sorted = INTEGER(sorted);
 
-  R_len_t n_subscript = Rf_length(sorted);
-  R_len_t n_missing = 0;
+  int extended[n_extend];
+  int i_extend = 0;
+  int nn = n;
 
-  qsort(p_sorted, n_subscript, sizeof(int), &qsort_icmp);
+  int* p_subscript = INTEGER(subscript);
 
+  R_len_t n_subscript = Rf_length(subscript);
   for (R_len_t i = 0; i < n_subscript; ++i) {
-    int elt = p_sorted[i];
+    int elt = p_subscript[i];
 
-    // All missing values are sorted to the beginning
-    if (elt == NA_INTEGER) {
-      ++n_missing;
+    // Missing value also covered here
+    if (elt <= n) {
       continue;
     }
-    --elt;
 
-    if (elt < n) {
-      continue;
+    // Special case: appending in ascending sequence at the end
+    // should not require any sorting
+    if (elt - 1 == nn) {
+      ++nn;
+      --n_extend;
+    } else {
+      extended[i_extend++] = elt - 1;
```

Comment:
The `extended` array was not zero-initialised, and this is the only path in the loop where it is assigned a value. This means the sorted array below will contain random memory in some cases.

Reply:
I'm not sure. I'm using only the first `i_extend` entries of the array during the sort and the subsequent check. Added a comment.

Reply:
I see, you're right.